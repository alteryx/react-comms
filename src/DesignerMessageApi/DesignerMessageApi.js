/* eslint-disable no-underscore-dangle */
import * as callback from '../Utils/callback';

class DesignerMessageApi {
  constructor(ctx) {
    this.context = ctx;
    this._model = {};
    this.subscriptions = new Map();
    this.context.Gui.SetConfiguration = currentToolConfiguration => {
      if (this.subscriptions.has('SetConfiguration')) {
        this.subscriptions.get('SetConfiguration')(currentToolConfiguration);
      }
    };
    this.context.Gui.GetConfiguration = () => {
      callback.JsEvent(this.context, 'GetConfiguration', this._model);
    };
  }

  sendMessage = (type, payload) => {
    callback.JsEvent(this.context, type, payload);
  };

  subscribe = (messageType, cb) => {
    this.subscriptions.set(messageType, cb);
  };

  get languageCode() {
    return this.context.AlteryxLanguageCode;
  }

  set model(toolConfiguration) {
    this._model = toolConfiguration;
  }
}

export default DesignerMessageApi;
