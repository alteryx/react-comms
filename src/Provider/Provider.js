import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import AyxAppWrapper from '../Core/AyxAppWrapper';
import UiSdkContext from '../Context';
import { isValidMessageType } from '../Utils/communication';

const Provider = props => {
  const envelope = window.dataEnvelope ? window.dataEnvelope : {};
  const [dataEnvelope, updateDataEnvelope] = useState(envelope);

  useEffect(() => {
    const receiveMessageEnvelope = ({ data }) => {
      if (isValidMessageType(data.type)) {
        updateDataEnvelope(data.payload);
      }
    };
    window.addEventListener('message', receiveMessageEnvelope);

    return function removeEventListener() {
      window.removeEventListener('message', receiveMessageEnvelope);
    };
  }, []);

  const { darkMode = false, productTheme = {}, locale = 'en', model = {} } = dataEnvelope;
  const { messages } = props;
  return (
    <UiSdkContext.Provider model={model}>
      <AyxAppWrapper
        id="app-wrapper"
        locale={locale}
        messages={messages}
        paletteType={darkMode ? 'dark' : 'light'}
        productTheme={productTheme}
      >
        {props.children}
      </AyxAppWrapper>
    </UiSdkContext.Provider>
  );
};

Provider.propTypes = {
  messages: PropTypes.objectOf(PropTypes.objectOf(PropTypes.string))
};

Provider.defaultProps = {
  messages: {}
};

export default Provider;
