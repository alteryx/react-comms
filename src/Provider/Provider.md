The provider is pretty dope. It provides lot of great info to your react app. 

```js { "file": "../demo.js" }
```