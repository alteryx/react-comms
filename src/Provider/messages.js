const messages = {
  en: {
    'example.label': 'super awesome string'
  },
  fr: {
    'example.label': 'ficelle super génial'
  },
  de: {
    'example.label': 'super tolle Saite'
  },
  es: {
    'example.label': 'cadena super impresionante'
  },
  pt: {
    'example.label': 'corda super incrível'
  },
  ja: {
    'example.label': '超素晴らしい文字列'
  },
  zh: {
    'example.label': '超棒的弦'
  }
};

export default messages;
