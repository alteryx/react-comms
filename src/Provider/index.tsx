/* eslint-disable no-underscore-dangle */
/* eslint-disable react/forbid-prop-types */
import React, { useEffect, useState } from 'react';

import { IAyxAppContext } from '../DesignerMessageApi';
import UiSdkContext, { IContextProviderProps } from '../Context';

interface IProviderProps {
  messages: object;
  messageBroker: {
    ayxAppContext: IAyxAppContext;
    model: object;
    subscribe: Function;
    sendMessage: Function;
  };
  children: React.ReactElement;
}

const Provider: React.FC = (props: IProviderProps) => {
  const { messages = {}, messageBroker } = props;
  const [model, updateModel] = useState(messageBroker.model);
  const [appContext, updateAppContext] = useState(messageBroker.ayxAppContext);

  const handleUpdateModel = newModel => {
    // The reason all 3 of these are here is to work in all use cases for now, DesignerMessageAPI and DevHarness.
    // TODO: Refactor this to only be dependent on one call
    updateModel(newModel);
    messageBroker.model = newModel;
    messageBroker.sendMessage('UPDATE_MODEL', newModel);
  };

  useEffect(() => {
    const receiveAppContext = data => {
      updateAppContext({ ...data });
    };
    const receiveModel = data => {
      updateModel({ ...data });
    };

    messageBroker.subscribe('UPDATE_MODEL', receiveModel);
    messageBroker.subscribe('UPDATE_APP_CONTEXT', receiveAppContext);
    return function cleanUp() {
      handleUpdateModel(messageBroker.model);
    };
  }, []);

  const providerProps: IContextProviderProps = {
    id: 'sdk-provider',
    value: [model, handleUpdateModel]
  };

  const { darkMode, locale, productTheme } = appContext || {};
  const appPropsToSpread = { messages, paletteType: darkMode ? 'dark' : 'light', theme: productTheme, locale };

  return (
    <UiSdkContext.Provider {...providerProps}>
      {React.cloneElement(props.children, { ...appPropsToSpread })}
    </UiSdkContext.Provider>
  );
};

export default Provider;
