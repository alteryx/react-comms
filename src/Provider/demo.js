/* eslint-disable no-unused-expressions */
import React from 'react';

import MessageApi from '../MessageApiBase/MessageApiBase';

import Provider from './Provider';

const messageBroker = new MessageApi(window);

const ProviderDemo = () => {
  console.log('thing');
  return <Provider messageBroker={messageBroker}>Hello world</Provider>;
};

<ProviderDemo />;
