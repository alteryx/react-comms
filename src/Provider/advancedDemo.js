import React from 'react';

import MessageApi from '../MessageApiBase/MessageApiBase.ts';
import UiSdkContext from '../Context/index.tsx';
import { Button, Grid, Input, FormControl, FormHelperText, FormControlLabel, Checkbox, InputLabel } from '../Core';

import Provider from './Provider.tsx';
import messages from './messages';

const initialModelState = {
  name: '',
  values: {
    field1: '',
    field2: '',
    checked: false
  }
};
// Instantiate your message handler to pass to the provider
const messageBroker = new MessageApi(window);
// In the real world, this line won't be here. Your model will take the shape of the parent application.
messageBroker.model = initialModelState;

const ProviderDemo = () => {
  const Child = () => {
    const [model, handleUpdateModel] = React.useContext(UiSdkContext);

    const resetModel = () => {
      handleUpdateModel(initialModelState);
    };

    const submitModel = () => {
      alert(
        `This is the message that would be getting sent to your parent app: ${JSON.stringify({
          type: 'UPDATE_MODEL',
          payload: model
        })}`
      );
    };

    const handleNameChange = event => {
      const newModel = { ...model, name: event.target.value };
      handleUpdateModel(newModel);
    };

    const handleValueChange = (prop, event) => {
      const values = { ...model.values, [prop]: event.target.value };
      const newModel = { ...model, values };
      handleUpdateModel(newModel);
    };

    const handleCheckedChange = () => {
      const values = { ...model.values, checked: !model.values.checked };
      const newModel = { ...model, values };
      handleUpdateModel(newModel);
    };

    return (
      <Grid container spacing={3}>
        <Grid item md={3} sm={6} xs={12}>
          <FormControl>
            <InputLabel htmlFor="component-simple">Name</InputLabel>
            <Input id="component-simple" onChange={handleNameChange} value={model.name} />
          </FormControl>
        </Grid>
        <Grid item md={3} sm={6} xs={12}>
          <FormControl aria-describedby="component-helper-text">
            <InputLabel htmlFor="component-helper">Value 1</InputLabel>
            <Input id="component-helper" onChange={e => handleValueChange('field1', e)} value={model.values.field1} />
            <FormHelperText id="component-helper-text">Some important helper text</FormHelperText>
          </FormControl>
        </Grid>
        <Grid item md={3} sm={6} xs={12}>
          <FormControl>
            <InputLabel htmlFor="component">Value 2</InputLabel>
            <Input id="component" onChange={e => handleValueChange('field2', e)} value={model.values.field2} />
            <FormHelperText>Some more important helper text</FormHelperText>
          </FormControl>
        </Grid>
        <Grid item>
          <FormControl component="fieldset">
            <FormControlLabel
              checked={model.values.checked}
              control={<Checkbox />}
              label="Checked"
              labelPlacement="top"
              onChange={handleCheckedChange}
              value="checked"
            />
          </FormControl>
        </Grid>
        <Grid item xs={3}>
          <Button color="primary" onClick={submitModel} variant="contained">
            Submit Model
          </Button>
        </Grid>
        <Grid item xs={3}>
          <Button onClick={resetModel} variant="contained">
            Reset Model
          </Button>
        </Grid>
      </Grid>
    );
  };

  return (
    <Provider messageBroker={messageBroker} messages={messages}>
      <Child />
    </Provider>
  );
};

<ProviderDemo />;
