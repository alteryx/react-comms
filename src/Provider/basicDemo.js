import React from 'react';

import MessageApi from '../MessageApiBase/MessageApiBase.ts';
import UiSdkContext from '../Context/index.tsx';
import { Button, Grid, Typography } from '../Core';

import Provider from './Provider.tsx';
import messages from './messages';

// Instantiate your message handler to pass to the provider
const messageBroker = new MessageApi(window);
// In the real world, this line won't be here. Your model will take the shape of the parent application.
messageBroker.model = { count: 0 };

const ProviderDemo = () => {
  const Child = () => {
    const [model, handleUpdateModel] = React.useContext(UiSdkContext);
    const incrementCount = () => {
      const newModel = { ...model };
      newModel.count++;
      handleUpdateModel(newModel);
    };
    return (
      <Grid container>
        <Grid item>
          <Button id="child" onClick={incrementCount} variant="contained">
            <Typography>Click me to increment the count</Typography>
          </Button>
        </Grid>
        <Grid item xs={12}>
          <Typography variant="h1">{model.count}</Typography>
        </Grid>
      </Grid>
    );
  };

  return (
    <Provider messageBroker={messageBroker} messages={messages}>
      <Child />
    </Provider>
  );
};

<ProviderDemo />;
