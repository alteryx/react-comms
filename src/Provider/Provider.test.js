/* eslint-disable no-underscore-dangle */
import React from 'react';
import { shallow, mount } from 'enzyme';

import DesignerMessageApi from '../DesignerMessageApi/DesignerMessageApi';
import UiSdkContext from '../Context';

import Provider from './Provider';

describe('Provider', () => {
  window.Alteryx = {
    Gui: {
      SetConfiguration: jest.fn(),
      GetConfiguration: jest.fn()
    }
  };
  const designerMessageApi = new DesignerMessageApi(window.Alteryx);
  designerMessageApi.subscribe = jest.fn();
  beforeAll(() => {
    window.dataEnvelope = {
      darkMode: false,
      productTheme: {},
      locale: 'en',
      model: { count: 0 }
    };
  });
  it('should be able to render on the DOM', () => {
    const wrapper = shallow(<Provider messageBroker={designerMessageApi} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render AyxAppWrapper with appropriate props if there is model data', () => {
    const wrapper = shallow(<Provider messageBroker={designerMessageApi} />);
    expect(wrapper.find('#app-wrapper').props()).toEqual({
      paletteType: 'light',
      productTheme: {},
      locale: 'en',
      children: undefined,
      id: 'app-wrapper',
      messages: {}
    });
  });

  it('should set subscriptions on the messageBroker for SetConfiguration events after it mounts', () => {
    mount(<Provider messageBroker={designerMessageApi} />);
    expect(designerMessageApi.subscribe).toHaveBeenCalledTimes(2);
  });

  it('should render the sdk-provider with a model when an event is emitted', () => {
    const wrapper = shallow(<Provider messageBroker={designerMessageApi} />);
    const valueProp = wrapper.find('#sdk-provider').prop('value');
    expect(valueProp).toHaveLength(2);
    expect(valueProp[0]).toEqual({});
    expect(valueProp[1]).toBeInstanceOf(Function);
  });

  it('should use the context hook to update the providers model', () => {
    const Child = () => {
      const [model, handleUpdateModel] = React.useContext(UiSdkContext);
      if (model.value !== 'foo') handleUpdateModel({ value: 'foo' });
      return <div id="child">{model.value}</div>;
    };
    const wrapper = mount(
      <Provider messageBroker={designerMessageApi}>
        <Child />
      </Provider>
    );

    expect(wrapper.find('#child').text()).toEqual('foo');
  });
});
