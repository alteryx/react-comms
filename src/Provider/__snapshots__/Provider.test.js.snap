// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider should be able to render on the DOM 1`] = `
<ContextProvider
  id="sdk-provider"
  value={
    Array [
      Object {},
      [Function],
    ]
  }
>
  <ForwardRef
    id="app-wrapper"
    locale="en"
    messages={Object {}}
    paletteType="light"
    productTheme={Object {}}
  />
</ContextProvider>
`;
